standard/array.lua

```diff
@@ -435,4 +436,90 @@ function Array.min(array, compare)
 	return Array.minBy(array, function(x) return x end, compare)
 end
 
+--[[
+Finds the index of the element in an array satisfying a predicate. Returs 0
+if no element satisfies the predicate.
+
+Example:
+
+Array.findIndex({3, 5, 4, 6, 7}, function(x) return x % 2 == 0 end)
+-- returns 3
+]]
+function Array.findIndex(array, pred)
+	for ix, elem in ipairs(array) do
+		if pred(elem, ix) then
+			return ix
+		end
+	end
+	return 0
+end
+
+--[[
+Returns the unique element in an array. Returns nil if there is more than one
+distinct element, or if the array is empty.
+
+Example:
+
+Array.uniqueElement({4, 4, 4})
+-- Returns 4
+]]
+function Array.uniqueElement(elems)
+	local uniqueElem
+	for i, elem in ipairs(elems) do
+		if i ~= 1 and elem ~= uniqueElem then
+			return nil
+		end
+		uniqueElem = elem
+	end
+	return uniqueElem
+end
+
+--[[
+Groups adjacent elements of an array based on applying a transformation to the
+elements. The function returns an array of groups.
+
+The optional equals parameter specifies the equality relation of the
+transformed elements.
+
+Example:
+Array.groupAdjacentBy({2, 3, 5, 7, 14, 16}, function(x) return x % 2 end)
+-- returns {{2}, {3, 5, 7}, {14, 16}}
+]]
+function Array.groupAdjacentBy(array, f, equals)
+	equals = equals or Logic.deepEquals
+
+	local groups = {}
+	local currentKey
+	for index, elem in ipairs(array) do
+		local key = f(elem)
+		if index == 1 or not equals(key, currentKey) then
+			currentKey = key
+			table.insert(groups, {})
+		end
+		table.insert(groups[#groups], elem)
+	end
+
+	return groups
+end
+
+--[[
+Returns distinct elements of an array.
+
+Example:
+
+Array.distinct({4, 5, 4, 3})
+-- Returns {4, 5, 3}
+]]
+function Array.distinct(elements)
```

Comment:
Make this `unique` instead, renaming all relevant pieces of code

Reply:
I think for our audience, AKA our contributors, `unique` is much more friendly.

Reply:
How about
```lua
function Array.unique(elements)
    return Array.distinct(element)
end
```
(or the other way around)

Reply:
My personal philosophy regarding our standard library is that it should be clean and simple. Following that thinking, my answer to your suggestion is that doesn't work.

Reply:
all the collection libraries i'm familiar with use `distinct`

Reply:
C++, Python Pandas, etc use `unique`